task/backend/scheduler/treescheduler.go

```diff
@@ -0,0 +1,403 @@
+// btree, then releases the lock.  We do not have to readjust the time on delete, because, if the minimum task isn't
+// ready yet, the main loop just resets the timer and keeps going.
+type TreeScheduler struct {
+	sync.RWMutex
+	scheduled    *btree.BTree
+	nextTime     map[ID]ordering // we need this index so we can delete items from the scheduled
+	when         time.Time
+	executor     Executor
+	onErr        ErrorFunc
+	time         Time
+	timer        Timer
+	done         chan struct{}
+	workchans    []chan Item
+	wg           sync.WaitGroup
+	checkpointer SchedulableService
+
+	sm *SchedulerMetrics
+}
+
+type ExecutorFunc func(ctx context.Context, id ID, scheduledAt time.Time) error
+
+type ErrorFunc func(ctx context.Context, taskID ID, scheduledAt time.Time, err error)
+
+type treeSchedulerOptFunc func(t *TreeScheduler) error
+
+func WithOnErrorFn(fn ErrorFunc) treeSchedulerOptFunc {
+	return func(t *TreeScheduler) error {
+		t.onErr = fn
+		return nil
+	}
+}
+
+func WithMaxConcurrentWorkers(n int) treeSchedulerOptFunc {
+	return func(t *TreeScheduler) error {
+		t.workchans = make([]chan Item, n)
+		return nil
+	}
+}
+
+func WithTime(t Time) treeSchedulerOptFunc {
+	return func(sch *TreeScheduler) error {
+		sch.time = t
+		return nil
+	}
+}
+
+// Executor is any function that accepts an ID, a time, and a duration.
+func NewScheduler(executor Executor, checkpointer SchedulableService, opts ...treeSchedulerOptFunc) (*TreeScheduler, *SchedulerMetrics, error) {
+	s := &TreeScheduler{
+		executor:     executor,
+		scheduled:    btree.New(degreeBtreeScheduled),
+		nextTime:     map[ID]ordering{},
+		onErr:        func(_ context.Context, _ ID, _ time.Time, _ error) {},
+		time:         stdTime{},
+		done:         make(chan struct{}, 1),
+		checkpointer: checkpointer,
+	}
+
+	// apply options
+	for i := range opts {
+		if err := opts[i](s); err != nil {
+			return nil, nil, err
+		}
+	}
+	if s.workchans == nil {
+		s.workchans = make([]chan Item, defaultMaxWorkers)
+
+	}
+
+	s.wg.Add(len(s.workchans))
+	for i := 0; i < len(s.workchans); i++ {
+		s.workchans[i] = make(chan Item)
+		go s.work(i)
+	}
+
+	s.sm = NewSchedulerMetrics(s)
+	s.when = s.time.Now().Add(maxWaitTime)
+	s.timer = s.time.NewTimer(maxWaitTime)
+	if executor == nil {
+		return nil, nil, errors.New("Executor must be a non-nil function")
+	}
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+	schedulerLoop:
+		for {
+			select {
+			case <-s.done:
+				s.Lock()
+				s.timer.Stop()
+				// close workchans
+				for i := range s.workchans {
+					close(s.workchans[i])
+				}
+				s.Unlock()
+				return
+			case <-s.timer.C():
+				s.Lock()
+				min := s.scheduled.Min()
+				if min == nil { // grab a new item, because there could be a different item at the top of the queue
+					s.when = s.time.Now().Add(maxWaitTime)
+					s.timer.Reset(maxWaitTime) // we can reset without stop, because its fired.
```

Comment:
If we can't find an item we will wait an hour?

Reply:
If there is no item in the tree at all, it will wait an hour and them check again (honestly, we could set it as long as we want without problem).  As soon as an item is added to the tree, the timer is reset to the appropriate time, so it just waits how long it needs to add.

Reply:
I guess my question is why wouldn't we just not set a new timer? What value does setting it to 1 hour do for us?